api/api/accounts/serializers.py

```diff
@@ -25,27 +26,159 @@ class Meta:
+        elif user_type == 'Staff':
+            if staff and staff_data:
+                self._update_staff(staff, staff_data)
+            elif staff_data:
+                Staff.objects.create(user=instance, **staff_data)
+            elif student_data or faculty_data:
+                raise serializers.ValidationError('Provided data doesn\'t match with user type')
+
+        return instance
+
+    def _update_base_user(self, instance, validated_data):
+        updatable_fields = ['first_name', 'last_name',
+                            'gender', 'contact_no', 'picture_url']
+
+        admin_only_editable_fields = ['email', 'user_type']
+
+        updatable_fields = self._validate_permission_and_get_updatable_fields(
+            instance,
+            validated_data,
+            updatable_fields,
+            admin_only_editable_fields)
+
+        self._update_instance(instance, validated_data, updatable_fields)
+
+    def _update_student(self, instance, validated_data):
+        updatable_fields = ['hostel_address', 'bio']
+        admin_only_editable_fields = ['roll_no', 'batch', 'department']
+
+        updatable_fields = self._validate_permission_and_get_updatable_fields(
+            instance,
+            validated_data,
+            updatable_fields,
+            admin_only_editable_fields)
+
+        self._update_instance(instance, validated_data, updatable_fields)
+
+    def _update_faculty(self, instance, validated_data):
+        updatable_fields = ['title', 'department', 'designation']
+
+        self._update_instance(instance, validated_data, updatable_fields)
+
+    def _update_staff(self, instance, validated_data):
+        updatable_fields = ['department', 'designation']
+
+        self._update_instance(instance, validated_data, updatable_fields)
+
+    def _update_instance(self, instance, validated_data, updatable_fields):
+
+        for field in updatable_fields:
+            field_stored_value = getattr(instance, field)
+            field_new_value = validated_data.get(field, field_stored_value)
+            setattr(instance, field, field_new_value)
+
+        instance.save()
+
+    def _validate_permission_and_get_updatable_fields(self,
+                                                      instance,
+                                                      validated_data,
+                                                      updatable_fields,
+                                                      admin_only_editable_fields
+                                                      ):
+        request_user = self.context['request'].user
+
+        if not request_user.is_staff:
+            for field in admin_only_editable_fields:
+                field_stored_value = getattr(instance, field)
+                field_new_value = validated_data.get(field, None)
+
+                if field_new_value and field_stored_value != field_new_value:
+                    raise serializers.ValidationError(f'{field} field is not editable')
```

Comment:
Again instead of throwing this error, let's just ignore those fields?

Avoid exceptions when they can be.